src/transaction/log_checkpoint_info.cpp

```diff
@@ -286,5 +287,164 @@ namespace cublog
+	  }
+	LSA_COPY (&tdes->head_lsa, &chkpt_one->head_lsa);
+	LSA_COPY (&tdes->tail_lsa, &chkpt_one->tail_lsa);
+	LSA_COPY (&tdes->undo_nxlsa, &chkpt_one->undo_nxlsa);
+	LSA_COPY (&tdes->posp_nxlsa, &chkpt_one->posp_nxlsa);
+	LSA_COPY (&tdes->savept_lsa, &chkpt_one->savept_lsa);
+	LSA_COPY (&tdes->tail_topresult_lsa, &chkpt_one->tail_topresult_lsa);
+	LSA_COPY (&tdes->rcv.tran_start_postpone_lsa, &chkpt_one->start_postpone_lsa);
+	tdes->client.set_system_internal_with_user (chkpt_one->user_name);
+	if (LOG_ISTRAN_2PC (tdes))
+	  {
+	    m_has_2pc = true;
+	  }
+      }
+
+
+    if (area != NULL)
+      {
+	free_and_init (area);
+      }
+
+    /*
+     * Now add top system operations that were in the process of
+     * commit to this transactions
+     */
+
+    log_page_local = (LOG_PAGE *) PTR_ALIGN (log_page_buffer, MAX_ALIGNMENT);
+    log_page_local->hdr.logical_pageid = NULL_PAGEID;
+    log_page_local->hdr.offset = NULL_OFFSET;
+
+    for (auto sysop : m_sysops)
+      {
+	chkpt_topone = &sysop;
+	tdes = logtb_rv_find_allocate_tran_index (thread_p, chkpt_topone->trid, &NULL_LSA);
+	if (tdes == NULL)
+	  {
+	    if (area != NULL)
+	      {
+		free_and_init (area);
+	      }
+
+	    logpb_fatal_error (thread_p, true, ARG_FILE_LINE, "log_recovery_analysis");
+	    return;
+	  }
+
+	if (tdes->topops.max == 0 || (tdes->topops.last + 1) >= tdes->topops.max)
+	  {
+	    if (logtb_realloc_topops_stack (tdes, m_sysops.size()) == NULL)
+	      {
+		if (area != NULL)
+		  {
+		    free_and_init (area);
+		  }
+
+		logpb_fatal_error (thread_p, true, ARG_FILE_LINE, "log_recovery_analysis");
+		return;
+	      }
+	  }
+
+	if (tdes->topops.last == -1)
+	  {
+	    tdes->topops.last++;
+	  }
+	else
+	  {
+	    assert (tdes->topops.last == 0);
+	  }
+	tdes->rcv.sysop_start_postpone_lsa = chkpt_topone->sysop_start_postpone_lsa;
+	tdes->rcv.atomic_sysop_start_lsa = chkpt_topone->atomic_sysop_start_lsa;
+	log_lsa_local = chkpt_topone->sysop_start_postpone_lsa;
+	error_code =
+		log_read_sysop_start_postpone (thread_p, &log_lsa_local, log_page_local, false, &sysop_start_postpone,
+					       NULL, NULL, NULL, NULL);
+	if (error_code != NO_ERROR)
+	  {
+	    assert (false);
+	    return;
+	  }
+	tdes->topops.stack[tdes->topops.last].lastparent_lsa = sysop_start_postpone.sysop_end.lastparent_lsa;
+	tdes->topops.stack[tdes->topops.last].posp_lsa = sysop_start_postpone.posp_lsa;
+      }
+  }
+  void
+  checkpoint_info::recovery_2pc_analysis (THREAD_ENTRY *thread_p) const
+  {
+    if (m_has_2pc)
+      {
+	return;
+      }
+
+    int tran_index;
+    LOG_TDES *tdes;		/* Transaction descriptor */
+
+    for (auto chkpt : m_trans)
+      {
+	tran_index = logtb_find_tran_index (thread_p, chkpt.trid);
+	if (tran_index != NULL_TRAN_INDEX)
+	  {
+	    tdes = LOG_FIND_TDES (tran_index);
+	    if (tdes != NULL && LOG_ISTRAN_2PC (tdes))
+	      {
+		log_2pc_recovery_analysis_info (thread_p, tdes, &chkpt.tail_lsa);
+	      }
+	  }
+      }
```

Comment:
Prefer to declare variables in the scope they're used.
```suggestion
    for (auto chkpt : m_trans)
      {
	int tran_index = logtb_find_tran_index (thread_p, chkpt.trid);
	if (tran_index != NULL_TRAN_INDEX)
	  {
	    LOG_TDES *tdes = LOG_FIND_TDES (tran_index);
	    if (tdes != NULL && LOG_ISTRAN_2PC (tdes))
	      {
		log_2pc_recovery_analysis_info (thread_p, tdes, &chkpt.tail_lsa);
	      }
	  }
      }
```